Add reference capability matrix section

EpicEric opened the pull request:
Closes #331

EpicEric commented:
I feel that this section got messier than I had planned. I still want to revisit the text before I feel comfortable merging this.

EpicEric commented on content/reference-capabilities/capability-matrix.md:
```diff
@@ -0,0 +1,53 @@
+* __Sendable capabilities__. If we want to send references to a different actor, we must make sure that the global and local aliases make the same guarantees. It'd be unsafe to send a `trn` to another actor, since we could possibly hold `box` references locally. Only `iso`, `val`, and `tag` have the same global and local restrictions – all of which are in the main diagonal of the matrix.
+* __Ephemeral subtyping__. If we have an ephemeral capability (for instance, `iso^` after consuming an isolated variable), we can be more permissive for the new alias, i.e. remove restrictions, such as allowing local aliases with read capabilities, and receive the reference into a `trn^`; or both read and write, which gives us `ref`. The same is true for more global alias, and we can get `val`, ` ...
+* __Recovering capabilities__. This is when we "lift" a capability, from a mutable reference to `iso` or an immutable reference to `val`. The matrix equivalent would be walking upwards starting from the capability – quite literally lifting in this case.
+* __Aliasing__. This is a bit less intuitive than the other points, since it only applies to `iso!` (which aliases to `tag`), and `trn!` (which aliases to `box`). If you picture a mirror in the anti-diagonal of the matrix (on top of `ref` and `val`), we can imagine that `iso` would be reflected on top of `tag` and `trn` would be reflected on top of `box`. Why? Well, our guarantees must be respect ...
```
I'm not particularly fond of this item, since it doesn't actually give much visual intuition, and might drop it altogether.

SeanTAllen commented:
Ping @jemc and I when you need a final review of this.

EpicEric commented:
Still might need adjustments, but it is ready for review.

@SeanTAllen @jemc

jemc commented on content/reference-capabilities/capability-matrix.md:
```diff
@@ -0,0 +1,55 @@
+
+Each reference capability in Pony is actually a pair of local guarantees and global guarantees. For instance, `ref` doesn't deny any read/write capabilities inside the actor, but denies other actors from reading or writing to that reference.
+
+You may recall from the [Reference Capability Guarantees](guarantees.md) section that mutable references cannot be safely shared between actors, while immutable references can be read by multiple actors. In other words, this implies that global properties are as restrictive or more restrictive than the local properties to that reference; it's not possible to write to an immutable reference in a l ...
```
This paragraph still seems a bit scattered to me. I've tried to suggest some changes that make it a bit more digestible, in my opinion:

```suggestion
You may recall from the [Reference Capability Guarantees](guarantees.md) section that mutable references cannot be safely shared between actors, while immutable references can be read by multiple actors. In general, global properties are always as restrictive or more restrictive than the local properties to that reference - what is denied globally must also be denied locally. For example, it's not possible to write to an immutable reference in either a global or local alias.
```

EpicEric replied:
I have added your change, but I'll review this paragraph myself later.

EpicEric commented:
@jemc I've added a few sentences to the paragraph you've mentioned, to try and make its conclusion more clear.

I have also squashed and rebased the branch.

jemc commented:
> I have also squashed and rebased the branch.

But now how will I know what the new sentences are? I hate squashing during review :disappointed_relieved: - in the future we can just squash on merge.

Can you point out what the changes were so I don't have to hunt and peck for them?

EpicEric commented:
I simply included two sentences to the end of your suggestion:

```diff
- [snip] For example, it's not possible to write to an immutable reference in either a global or local alias.
+ [snip] For example, it's not possible to write to an immutable reference in either a global or local alias. It's also not possible to read from or write to an opaque reference, `tag`. Therefore, some combinations of local and global aliases are impossible, and have no designated capabilities.
```
